README.rst
GS: A minimalistic Google Storage client
========================================

*gs* is a command line interface (CLI) and Python library that provides a set of essential commands for
`Google Cloud Storage <https://cloud.google.com/storage/>`_. It is modeled after the AWS CLI's ``aws s3`` command. Its
features are:

* Python 3 compatibility
* A minimalistic set of dependencies
* A tiny footprint
* Intuitive convention-driven configuration of API credentials without browser login prompts
* Checksum validation to ensure end-to-end data integrity in uploads and downloads
* Progress bars for long-running upload and download operations
* Resumable uploads and downloads
* Multithreaded directory sync and batch delete, capable of handling large numbers of objects
* An attractive paging and table layout interface
* A JSON object metadata output mode for feeding data to other utilities

Installation
~~~~~~~~~~~~
::

   pip install gs

Synopsis
~~~~~~~~
Usage:
  ``gs [OPTIONS] COMMAND [ARGS]...``

Options:
  --version  Show the version and exit.
  --help     Show this message and exit.

Commands:
   +------------------+--------------------------------------------------+
   | ``gs configure`` | Set gs config options, including the API key.    |
   +------------------+--------------------------------------------------+
   | ``gs ls``        | List buckets or objects in a bucket/prefix.      |
   +------------------+--------------------------------------------------+
   | ``gs cp``        | Copy files to, from, or between buckets.         |
   +------------------+--------------------------------------------------+
   | ``gs mv``        | Move files to, from, or between buckets.         |
   +------------------+--------------------------------------------------+
   | ``gs mb``        | Create a new Google Storage bucket.              |
   +------------------+--------------------------------------------------+
   | ``gs rb``        | Permanently delete an empty bucket.              |
   +------------------+--------------------------------------------------+
   | ``gs rm``        | Delete objects (files) from buckets.             |
   +------------------+--------------------------------------------------+
   | ``gs sync``      | Sync a directory of files with bucket/prefix.    |
   +------------------+--------------------------------------------------+
   | ``gs api``       | Use httpie to perform a raw HTTP API request.    |
   +------------------+--------------------------------------------------+
   | ``gs presign``   | Get a pre-signed URL for accessing an object.    |
   +------------------+--------------------------------------------------+

Run ``gs configure`` to configure Google service account access credentials that will be used by the
``gs`` command. You can create a new service account key at https://console.cloud.google.com/iam-admin/serviceaccounts.

Credentials
~~~~~~~~~~~
Before making API calls, *gs* ingests API credentials in the following order of priority:

- First, *gs* checks if a ``GOOGLE_APPLICATION_CREDENTIALS`` environment variable is set. If so, it attempts to load and use
  credentials from the service account credentials filename referenced by the variable.
- If that varible is not set, *gs* attempts to load service account credentials previously configured with ``gs configure``
  (stored in ``~/.config/gs/config.json``).
- If that fails, *gs* attempts to load a service account API token from
  `Google instance metadata <https://cloud.google.com/compute/docs/storing-retrieving-metadata>`_.
- If that fails, *gs* prints a warning and attempts to make API requests
  `anonymously <https://cloud.google.com/storage/docs/access-public-data>`_.

Using the Python library interface
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.. code-block:: python

    from gs import GSClient
    client = GSClient()
    object_meta = client.get("b/my-bucket/o/my-object")
    with client.get("b/my-bucket/o/my-object", params=dict(alt="media"), stream=True) as res:
        object_bytes = res.raw.read()
    presigned_url = client.get_presigned_url("my-bucket", "my-object", expires_at=time.time()+3600)

Authors
-------
* Andrey Kislyuk

Links
-----
* `Project home page (GitHub) <https://github.com/chanzuckerberg/gs>`_
* `Package distribution (PyPI) <https://pypi.python.org/pypi/gs>`_
* `Change log <https://github.com/chanzuckerberg/gs/blob/master/Changes.rst>`_
* `GCB builds <https://console.cloud.google.com/cloud-build/builds>`_
* `Google Cloud Storage <https://cloud.google.com/storage/>`_

Bugs
~~~~
Please report bugs, issues, feature requests, etc. on `GitHub <https://github.com/chanzuckerberg/gs/issues>`_.

License
-------
Licensed under the terms of the MIT License.

.. image:: https://travis-ci.com/chanzuckerberg/gs.png
   :target: https://travis-ci.com/chanzuckerberg/gs
.. image:: https://img.shields.io/pypi/v/gs.svg
   :target: https://pypi.python.org/pypi/gs
.. image:: https://img.shields.io/pypi/l/gs.svg
   :target: https://pypi.python.org/pypi/gs
.. image:: https://codecov.io/gh/chanzuckerberg/gs/branch/master/graph/badge.svg
   :target: https://codecov.io/gh/chanzuckerberg/gs


Changes.rst
Changes for v1.0.0 (2020-07-15)
===============================

-  Bump crc32c version

-  Support anonymous requests

-  Package is stable

Changes for v0.5.6 (2019-02-15)
===============================

-  gs cp: support trailing wildcards

-  gs rm: support dryrun parameter

Changes for v0.5.5 (2019-02-13)
===============================

-  Support globs in gs rm

Changes for v0.5.4 (2019-02-10)
===============================

-  Re-enable progressbars on Python 2

Changes for v0.5.3 (2019-02-10)
===============================

-  Correctly discard old partial uploads

Changes for v0.5.2 (2019-02-10)
===============================

-  Cosmetic fixups for batch

Changes for v0.5.1 (2019-02-09)
===============================

-  gs rm: Implement multithreaded batch delete

Changes for v0.5.0 (2019-02-09)
===============================

-  gs sync: use multithreaded upload and download; batch list remote
   objects

-  Bump and constrain dependencies

Changes for v0.4.4 (2019-01-24)
===============================

-  Require error response to be JSON to print body

Changes for v0.4.3 (2019-01-24)
===============================

-  Exit with non-zero status if error formatting fails

Changes for v0.4.2 (2019-01-24)
===============================

-  Handle no buckets condition

Changes for v0.4.1 (2019-01-24)
===============================

-  Format HTTP wire errors

Changes for v0.4.0 (2018-09-28)
===============================

-  Expand metadata handling

-  Simplify timeout setting

Changes for v0.3.9 (2018-09-28)
===============================

Bump crc32c dependency to v1.5

Changes for v0.3.8 (2018-09-28)
===============================

-  Avoid race conditions in get_session

Changes for v0.3.7 (2018-09-26)
===============================

-  Doc improvements

Changes for v0.3.6 (2018-09-26)
===============================

-  Add sanity checks to release script
-  Add doc for gs api

Changes for v0.3.5 (2018-09-26)
===============================

-  Remove dependency on prompt-toolkit

-  Begin gs api

-  Change license from Apache to MIT (#4)

Changes for v0.3.4 (2018-09-14)
===============================

-  Set retry and timeout policy on HTTP requests

Changes for v0.3.3 (2018-09-12)
===============================

Avoid map on python3

Changes for v0.3.2 (2018-09-12)
===============================

-  Avoid crashing when config save fails

Changes for v0.3.1 (2018-09-12)
===============================

-  Fix project discovery

-  Expand user paths in local filenames



test/test.py
#!/usr/bin/env python
# coding: utf-8

import os, sys, unittest, uuid, tempfile, time, logging

from gs.util.compat import TemporaryDirectory

import gs, tweak
from gs import cli

from gs.util.compat import USING_PYTHON2

logging.basicConfig(level=logging.DEBUG)

class TestGS(unittest.TestCase):
    test_bucket = os.environ["GS_TEST_BUCKET"]
    test_id = int(time.time() * 1000)

    def test_ls_commands(self):
        config = tweak.Config("gs", save_on_exit=False)
        client = gs.GSClient(config=config)
        cli.ls.main([], standalone_mode=False)
        cli.ls.main(["--json"], standalone_mode=False)

    def test_rw_commands(self):
        with tempfile.NamedTemporaryFile() as tf1, tempfile.NamedTemporaryFile() as tf2:
            payload = os.urandom(1024 * 1024 + 1)
            tf1.write(payload)
            tf1.flush()
            tf2.write(payload[:16])
            tf2.flush()
            test_prefix = "gs://{}/{}".format(self.test_bucket, self.test_id)
            furl1 = os.path.join(test_prefix, os.path.basename(tf1.name))
            furl2 = os.path.join(test_prefix, os.path.basename(tf2.name))
            cli.cp.main([tf1.name, tf2.name, test_prefix], standalone_mode=False)
            cli.ls.main([test_prefix], standalone_mode=False)
            # cli.presign.main([furl1], standalone_mode=False)
            with TemporaryDirectory() as td:
                cli.cp.main([furl1, furl2, td], standalone_mode=False)
                with open(os.path.join(td, tf1.name), "rb") as fh:
                    self.assertEqual(fh.read(), payload)
                with open(os.path.join(td, tf2.name), "rb") as fh:
                    self.assertEqual(fh.read(), payload[:16])
                cli.cp.main([furl1, furl1 + ".2"], standalone_mode=False)
                cli.mv.main([furl1, furl1 + ".3"], standalone_mode=False)
                cli.ls.main([furl1 + ".*"], standalone_mode=False)
                cli.cp.main([furl1 + ".*", furl1 + "-dir"], standalone_mode=False)
                cli.cp.main([furl1 + ".3", furl1 + ".4"], standalone_mode=False)
                cli.rm.main([furl1 + ".2", furl1 + ".3"], standalone_mode=False)
                cli.sync.main([test_prefix, td], standalone_mode=False)
                cli.sync.main([td, test_prefix], standalone_mode=False)
                cli.rm.main([test_prefix, "--dryrun", "--recursive"], standalone_mode=False)
                cli.rm.main([test_prefix, "--recursive"], standalone_mode=False)

if __name__ == "__main__":
    unittest.main()


gs/util/compat.py
from __future__ import absolute_import, division, print_function, unicode_literals

import os, sys, datetime, errno, threading

USING_PYTHON2 = True if sys.version_info < (3, 0) else False

if USING_PYTHON2:
    from multiprocessing import cpu_count
    from thread import get_ident
    from StringIO import StringIO
    from repr import Repr
    str = unicode # noqa
    from ..packages.backports.functools_lru_cache import lru_cache
    from ..packages.backports.shutil_get_terminal_size import get_terminal_size
    from ..packages.backports.tempfile import TemporaryDirectory

    def makedirs(name, mode=0o777, exist_ok=False):
        try:
            os.makedirs(name, mode)
        except OSError as e:
            if not (exist_ok and e.errno == errno.EEXIST and os.path.isdir(name)):
                raise

    def median(data):
        data = sorted(data)
        n = len(data)
        if n == 0:
            raise Exception("no median for empty data")
        if n % 2 == 1:
            return data[n // 2]
        else:
            i = n // 2
            return (data[i - 1] + data[i]) / 2

    def timestamp(dt):
        if dt.tzinfo is None:
            from time import mktime
            return mktime((dt.year, dt.month, dt.day, dt.hour, dt.minute, dt.second, -1, -1, -1)) + dt.microsecond / 1e6
        else:
            from dateutil.tz import tzutc
            return (dt - datetime.datetime(1970, 1, 1, tzinfo=tzutc())).total_seconds()

    def thread_is_main():
        return True if threading.current_thread().name == "MainThread" else False
else:
    from threading import get_ident
    from io import StringIO
    from reprlib import Repr
    str = str
    from functools import lru_cache
    from shutil import get_terminal_size
    from tempfile import TemporaryDirectory
    from os import makedirs, cpu_count
    from statistics import median
    timestamp = datetime.datetime.timestamp

    def thread_is_main():
        return True if threading.current_thread() is threading.main_thread() else False


gs/util/__init__.py
import os, sys, struct, warnings, functools
from datetime import datetime

from dateutil.parser import parse as dateutil_parse
from dateutil.relativedelta import relativedelta
import requests.exceptions

from .compat import USING_PYTHON2

class Timestamp(datetime):
    """
    Integer inputs are interpreted as milliseconds since the epoch. Sub-second precision is discarded. Suffixes (s, m,
    h, d, w) are supported. Negative inputs (e.g. -5m) are interpreted as relative to the current date. Other inputs
    (e.g. 2020-01-01, 15:20) are parsed using the dateutil parser.
    """
    def __new__(cls, t):
        if isinstance(t, (str, bytes)) and t.isdigit():
            t = int(t)
        if not isinstance(t, (str, bytes)):
            from dateutil.tz import tzutc
            return datetime.fromtimestamp(t // 1000, tz=tzutc())
        try:
            units = {"weeks", "days", "hours", "minutes", "seconds"}
            diffs = {u: float(t[:-1]) for u in units if u.startswith(t[-1])}
            if len(diffs) == 1:
                return datetime.now().replace(microsecond=0) + relativedelta(**diffs)
            return dateutil_parse(t)
        except (ValueError, OverflowError, AssertionError):
            raise ValueError('Could not parse "{}" as a timestamp or time delta'.format(t))

class CRC32C:
    def __init__(self, data=None):
        import crc32c
        self._crc32c = crc32c
        self._csum = crc32c.crc32(data if data is not None else b"")

    def update(self, data):
        self._csum = self._crc32c.crc32(data, self._csum)

    def digest(self):
        return long_to_bytes(self._csum)

def long_to_bytes(n):
    s = b''
    if USING_PYTHON2:
        n = long(n)  # noqa
    pack = struct.pack
    while n > 0:
        s = pack(b'>I', n & 0xffffffff) + s
        n = n >> 32
    for i in range(len(s)):
        if s[i] != b'\000'[0]:
            break
    else:
        s = b'\000'
        i = 0
    s = s[i:]
    return s

def get_file_size(path):
    try:
        return os.path.getsize(path)
    except Exception:
        return -1

def format_http_errors(fn):
    @functools.wraps(fn)
    def error_formatter(*args, **kwargs):
        try:
            return fn(*args, **kwargs)
        except requests.exceptions.HTTPError as e:
            try:
                e.response.json()
                msg = e.response.content.decode()
            except Exception:
                msg = "{}: {}".format(type(e).__name__, e)
            exit(msg)
    return error_formatter

def batches(iterable, batch_size=None):
    batch = []
    for i in iterable:
        batch.append(i)
        if len(batch) >= batch_size:
            yield batch
            batch = []
    if len(batch) > 0:
        yield batch


gs/util/printing.py
# coding: utf-8
from __future__ import absolute_import, division, print_function, unicode_literals

import os, sys, json, shutil, subprocess, re, errno, threading
from datetime import datetime, timedelta
import click
from .exceptions import GetFieldError, GSException
from .compat import str, get_terminal_size, thread_is_main

USING_PYTHON2 = True if sys.version_info < (3, 0) else False

ANSI_OK = True if sys.stdout.isatty() and os.name != "nt" else False

def CYAN(message=None):
    if message is None:
        return "\033[36m" if ANSI_OK else ""
    else:
        return CYAN() + message + ENDC()

def BLUE(message=None):
    if message is None:
        return "\033[34m" if ANSI_OK else ""
    else:
        return BLUE() + message + ENDC()

def YELLOW(message=None):
    if message is None:
        return "\033[33m" if ANSI_OK else ""
    else:
        return YELLOW() + message + ENDC()

def GREEN(message=None):
    if message is None:
        return "\033[32m" if ANSI_OK else ""
    else:
        return GREEN() + message + ENDC()

def RED(message=None):
    if message is None:
        return "\033[31m" if ANSI_OK else ""
    else:
        return RED() + message + ENDC()

def WHITE(message=None):
    if message is None:
        return "\033[37m" if ANSI_OK else ""
    else:
        return WHITE() + message + ENDC()

def UNDERLINE(message=None):
    if message is None:
        return "\033[4m" if ANSI_OK else ""
    else:
        return UNDERLINE() + message + ENDC()

def BOLD(message=None):
    if message is None:
        return "\033[1m" if ANSI_OK else ""
    else:
        return BOLD() + message + ENDC()

def ENDC():
    return "\033[0m" if ANSI_OK else ""

def border(i):
    return WHITE() + i + ENDC()

ansi_pattern = re.compile(r"(\x9B|\x1B\[)[0-?]*[ -\/]*[@-~]")

def strip_ansi_codes(i):
    return re.sub(ansi_pattern, "", i)

def ansi_truncate(s, max_len):
    ansi_total_len = 0
    for ansi_code in ansi_pattern.finditer(s):
        ansi_code_start, ansi_code_end = ansi_code.span()
        if ansi_code_start > max_len + ansi_total_len - 1:
            break
        ansi_total_len += ansi_code_end - ansi_code_start
    if len(s) > max_len + ansi_total_len:
        return s[:max_len + ansi_total_len - 1] + "…"
    return s

def format_table(table, column_names=None, column_specs=None, max_col_width=32, auto_col_width=False):
    """
    Table pretty printer. Expects tables to be given as arrays of arrays::

        print(format_table([[1, "2"], [3, "456"]], column_names=['A', 'B']))
    """
    orig_col_args = dict(column_names=column_names, column_specs=column_specs)
    if len(table) > 0:
        col_widths = [0] * len(table[0])
    elif column_specs is not None:
        col_widths = [0] * (len(column_specs) + 1)
    elif column_names is not None:
        col_widths = [0] * len(column_names)
    my_col_names = []
    if column_specs is not None:
        column_names = ["Row"]
        column_names.extend([col["name"] for col in column_specs])
        column_specs = [{"name": "Row", "type": "float"}] + column_specs
    if column_names is not None:
        for i in range(len(column_names)):
            my_col = ansi_truncate(str(column_names[i]), max_col_width)
            my_col_names.append(my_col)
            col_widths[i] = max(col_widths[i], len(strip_ansi_codes(my_col)))
    trunc_table = []
    for row in table:
        my_row = []
        for i in range(len(row)):
            my_item = ansi_truncate(str(row[i]), max_col_width)
            my_row.append(my_item)
            col_widths[i] = max(col_widths[i], len(strip_ansi_codes(my_item)))
        trunc_table.append(my_row)

    type_colormap = {"boolean": BLUE(),
                     "integer": YELLOW(),
                     "float": WHITE(),
                     "string": GREEN()}
    for i in "uint8", "int16", "uint16", "int32", "uint32", "int64":
        type_colormap[i] = type_colormap["integer"]
    type_colormap["double"] = type_colormap["float"]

    def col_head(i):
        if column_specs is not None:
            return BOLD() + type_colormap[column_specs[i]["type"]] + column_names[i] + ENDC()
        else:
            return BOLD() + WHITE() + column_names[i] + ENDC()

    formatted_table = [border("┌") + border("┬").join(border("─") * i for i in col_widths) + border("┐")]
    if len(my_col_names) > 0:
        padded_column_names = [col_head(i) + " " * (col_widths[i] - len(my_col_names[i]))
                               for i in range(len(my_col_names))]
        formatted_table.append(border("│") + border("│").join(padded_column_names) + border("│"))
        formatted_table.append(border("├") + border("┼").join(border("─") * i for i in col_widths) + border("┤"))

    for row in trunc_table:
        padded_row = [row[i] + " " * (col_widths[i] - len(strip_ansi_codes(row[i]))) for i in range(len(row))]
        formatted_table.append(border("│") + border("│").join(padded_row) + border("│"))
    formatted_table.append(border("└") + border("┴").join(border("─") * i for i in col_widths) + border("┘"))

    if auto_col_width:
        if not sys.stdout.isatty():
            raise GSException("Cannot auto-format table, output is not a terminal")
        table_width = len(strip_ansi_codes(formatted_table[0]))
        tty_cols, tty_rows = get_terminal_size()
        if table_width > max(tty_cols, 80):
            return format_table(table, max_col_width=max_col_width - 1, auto_col_width=True, **orig_col_args)
    return "\n".join(formatted_table)

def page_output(content, pager=None, file=None):
    if file is None:
        file = sys.stdout
    if not content.endswith("\n"):
        content += "\n"

    pager_process = None
    try:
        if file != sys.stdout or not file.isatty() or not content.startswith(border("┌")):
            raise GSException()
        content_lines = content.splitlines()
        content_rows = len(content_lines)

        tty_cols, tty_rows = get_terminal_size()

        naive_content_cols = max(len(i) for i in content_lines)
        if tty_rows > content_rows and tty_cols > naive_content_cols:
            raise GSException()

        content_cols = max(len(strip_ansi_codes(i)) for i in content_lines)
        if tty_rows > content_rows and tty_cols > content_cols:
            raise GSException()

        pager_process = subprocess.Popen(pager or os.environ.get("PAGER", "less -RS"), shell=True,
                                         stdin=subprocess.PIPE, stdout=file)
        pager_process.stdin.write(content.encode("utf-8"))
        pager_process.stdin.close()
        pager_process.wait()
        if pager_process.returncode != os.EX_OK:
            raise GSException()
    except Exception as e:
        if not (isinstance(e, IOError) and e.errno == errno.EPIPE):
            file.write(content.encode("utf-8") if USING_PYTHON2 else content)
    finally:
        try:
            pager_process.terminate()
        except BaseException:
            pass

def get_field(item, field):
    for element in field.split("."):
        try:
            item = getattr(item, element)
        except AttributeError:
            try:
                item = item.get(element)
            except AttributeError:
                raise GetFieldError('Unable to access field or attribute "{}" of {}'.format(field, item))
    return item

def format_datetime(d):
    from dateutil.tz import tzutc
    from babel import dates
    d = d.replace(microsecond=0)
    if not USING_PYTHON2:
        # Switch from UTC to local TZ
        d = d.astimezone(tz=None)
    return dates.format_timedelta(d - datetime.now(tzutc()), add_direction=True)

def format_cell(cell):
    if isinstance(cell, datetime):
        cell = format_datetime(cell)
    if isinstance(cell, timedelta):
        from babel import dates
        cell = dates.format_timedelta(-cell, add_direction=True)
    if isinstance(cell, (list, dict)):
        cell = json.dumps(cell, default=lambda x: str(x))
    return cell

def get_cell(resource, field, transform=None):
    cell = get_field(resource, field)
    if transform:
        try:
            cell = transform(cell, resource)
        except TypeError:
            cell = transform(cell)
    return ", ".join(i.name for i in cell.all()) if hasattr(cell, "all") else cell

def format_tags(cell, row):
    tags = {tag["Key"]: tag["Value"] for tag in cell} if cell else {}
    return ", ".join("{}={}".format(k, v) for k, v in tags.items())

def trim_names(names, *prefixes):
    for name in names:
        for prefix in prefixes:
            if name.startswith(prefix):
                name = name[len(prefix):]
        yield name

def format_number(n, fractional_digits=2):
    B = n
    KB = float(1024)
    MB = float(KB * 1024)
    GB = float(MB * 1024)
    TB = float(GB * 1024)

    if B < KB:
        return '{0}'.format(B)
    elif KB <= B < MB:
        return '{0:.{precision}f}K'.format(B / KB, precision=fractional_digits)
    elif MB <= B < GB:
        return '{0:.{precision}f}M'.format(B / MB, precision=fractional_digits)
    elif GB <= B < TB:
        return '{0:.{precision}f}G'.format(B / GB, precision=fractional_digits)
    elif TB <= B:
        return '{0:.{precision}f}T'.format(B / TB, precision=fractional_digits)

def tabulate(collection, args, cell_transforms=None):
    if cell_transforms is None:
        cell_transforms = {}
    cell_transforms["tags"] = format_tags
    if getattr(args, "json", None):
        table = [{f: get_cell(i, f, cell_transforms.get(f)) for f in args.columns} for i in collection]
        return json.dumps(table, indent=2, default=lambda x: str(x))
    else:
        table = [[get_cell(i, f, cell_transforms.get(f)) for f in args.columns] for i in collection]
        if getattr(args, "sort_by", None):
            reverse = False
            if args.sort_by.endswith(":reverse"):
                reverse = True
                args.sort_by = args.sort_by[:-len(":reverse")]
            table = sorted(table, key=lambda x: x[args.columns.index(args.sort_by)], reverse=reverse)
        table = [[format_cell(c) for c in row] for row in table]
        args.columns = list(trim_names(args.columns, *getattr(args, "trim_col_names", [])))
        format_args = dict(auto_col_width=True) if args.max_col_width == 0 else dict(max_col_width=args.max_col_width)
        return format_table(table, column_names=getattr(args, "display_column_names", args.columns), **format_args)

def get_progressbar(**kwargs):
    bar = click.progressbar(**kwargs)
    if not thread_is_main():
        bar.is_hidden = True
    return bar
